src/main/java/org/springframework/data/elasticsearch/repository/support/querybyexample/ExampleCriteriaMapper.java

```diff
@@ -0,0 +1,174 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Maps a {@link Example} to a {@link org.springframework.data.elasticsearch.core.query.Criteria}
+ *
+ * @param <T> Class type
+ * @author Ezequiel Antúnez Camacho
+ */
+class ExampleCriteriaMapper<T> {
+
+	private static final Set<ExampleMatcher.StringMatcher> SUPPORTED_MATCHERS = EnumSet.of(
+			ExampleMatcher.StringMatcher.DEFAULT, ExampleMatcher.StringMatcher.EXACT, ExampleMatcher.StringMatcher.STARTING,
+			ExampleMatcher.StringMatcher.CONTAINING, ExampleMatcher.StringMatcher.ENDING);
+
+	private final MappingContext<? extends ElasticsearchPersistentEntity<T>, ElasticsearchPersistentProperty> mappingContext;
+
+	/**
+	 * Builds a {@link ExampleCriteriaMapper}
+	 *
+	 * @param mappingContext mappingContext to use
+	 */
+	ExampleCriteriaMapper(
+			MappingContext<? extends ElasticsearchPersistentEntity<T>, ElasticsearchPersistentProperty> mappingContext) {
+		this.mappingContext = mappingContext;
+	}
+
+	<S extends T> Criteria criteria(Example<S> example) {
+		return buildCriteria(example);
+	}
+
+	private <S extends T> Criteria buildCriteria(Example<S> example) {
+		final ExampleMatcherAccessor matcherAccessor = new ExampleMatcherAccessor(example.getMatcher());
+
+		return applyPropertySpecs(new Criteria(), "", example.getProbe(),
+				mappingContext.getRequiredPersistentEntity(example.getProbeType()), matcherAccessor,
+				example.getMatcher().getMatchMode());
+	}
+
+	private Criteria applyPropertySpecs(Criteria criteria, String path, @Nullable Object probe,
+			ElasticsearchPersistentEntity<?> persistentEntity, ExampleMatcherAccessor exampleSpecAccessor,
+			ExampleMatcher.MatchMode matchMode) {
+
+		if (probe == null) {
+			return criteria;
+		}
+
+		PersistentPropertyAccessor<?> propertyAccessor = persistentEntity.getPropertyAccessor(probe);
+
+		for (ElasticsearchPersistentProperty property : persistentEntity) {
+			final String propertyName = getPropertyName(property);
+			String propertyPath = StringUtils.hasText(path) ? (path + "." + propertyName) : propertyName;
+			if (exampleSpecAccessor.isIgnoredPath(propertyPath) || property.isCollectionLike()
+					|| property.isVersionProperty()) {
+				continue;
+			}
+
+			Object propertyValue = propertyAccessor.getProperty(property);
+			if (property.isMap() && propertyValue != null) {
+				for (Map.Entry<String, Object> entry : ((Map<String, Object>) propertyValue).entrySet()) {
+					String key = entry.getKey();
+					Object value = entry.getValue();
+					criteria = applyPropertySpec(propertyPath + "." + key, value, exampleSpecAccessor, property, matchMode,
+							criteria);
+				}
+				continue;
+			}
+
+			criteria = applyPropertySpec(propertyPath, propertyValue, exampleSpecAccessor, property, matchMode, criteria);
+		}
+		return criteria;
+	}
+
+	private String getPropertyName(ElasticsearchPersistentProperty property) {
+		return property.isIdProperty() ? "_id" : property.getName();
```

Comment:
I don't think this is correct. When working with `Criteria` the property names are the names of the Java entity class properties.
The mapping of the names to potentially different names in Elasticsearch and the check which of these properties is used as id is done, when the actual query is built from the `Criteria`. Here it's all about the Java property names.

Reply:
Removed in https://github.com/spring-projects/spring-data-elasticsearch/pull/2422/commits/0b36a268b21075a71a4a3447c864fa3b968e6356